### ISSUE TITLE
C++ serialization

### ISSUE THREAD

**Issue by pavlis** [wangyinz/mspass#32]
A nasty feature created by the changes in Metadata I implemented for mspass created a problem with serialization.   As you know I used boost's serialization library extensively for the seispp stuff in Antelope.   You may not recall that I found the change to boost::any to make Metadata more like a python dictionary (allows mixed types) broke boost's serialization. The "boost::any" stuff I used won't serialize, which is not actually very surprising as it very much complicates serialization to allow mixed types in an std::map container.  

I raise this now because working with yaml-cpp suggests that a clean way to handle serialization of the C++ data objects may be to use yaml for all or part of the data serialization.   The documentation for yaml-cpp is pretty weak, but suggests this is quite possible.   To make it efficient, however, I would bet a fair amount you would only want to serialize Metadata with yaml.   

The obvious alternative is to crack the pybind11 method for serializing such things with pickle.   That would likely be a lot easier and probably how we should start.  I raise this now mostly to remind us if we run into an efficiency problem with pickle we may want to revisit this issue.   

I'm also a bit unclear what role serialization will play in a map/reduce operation.   It seems to me it only enters in either reading or writing data to the database or exporting/importing data.   i.e. if I read the spark documentation right a single processing script that contains a series of map/reduce operations on and rdd will handle the movement of data between processing steps automatically in memory or caching to disk.   Do I understand this correctly?

**Comment by wangyinz** [comment]
I agree that it is probably better to start with pickle for now. It is probably not very ideal for C++, so we do need to keep an eye on any good alternatives out there. 

Yes, Spark does its own serialization internally, and we don't need to worry about that.

**Comment by pavlis** [comment]
I just ran some tests with TimeSeries and Seismogram in python.   I discovered pickle won't work without some effort.  It requires yet another set of tricks with pybind11.   However, I discovered something I think might be better anyway.   

The wrappers I built for TimeSeries and Seismogram use a trick to make the sample data look like numpy arrays;  TimeSeries std::vector looks like a 1D numpy array to python and Seisogram dmatrix looks like a 3xns numpy matrix.   I found this sequence worked fine:
import numpy
from mspasspy import CoreTimeSeries
from mspasspy import CoreSeismogram
d=CoreTimeSeries(100)
# Note this won't work unless the file arg has the .npy extension 
numpy.save('test.npy',d.s)
x=numpy.load('test.npy')
print(x)   # should return a list of 100 zeros
d=CoreSeismogram(100)
numpy.save('testseis.npy',d.u)
x=numpy.load('testseis.npy')
print(x)   # should print a 3x100 matrix of zeros

Cut and paste that and you should find it works.   

Now the unknown is how this would work with a write to gridfs.   I haven't tested that yet, but suspect it is just a matter of getting the right incantation.  

This gets us very close to a database reader and writer for the C++ data objects IF we use the model of handling Metadata differently from the sample data.   The MongoDBConverter class I just put into master this morning is the interface needed for the Metadata component.  This is a (likely) solution for the sample data.   

Want to hear your response before I dive deeper into this.

**Comment by wangyinz** [comment]
I think it is definitely more efficient to use the .npy file than pickle to store data. I do have the concern that if the C++ codes do not support pickle, it might be confusing to the Python programmers as that is a very common quick and easy way to dump any data. I briefly looked at the pybind11 documents and see that there is quite a bit of work to put into to make pickle work, and it will only work for a certain version of pickle - I can totally see why you proposed this alternative.

Anyway, I think the .npy is an OK solution for now. Eventually, we will want to have pickle functioning too as an alternative for some users (it will be a slower but cleaner interface since one can dump everything all together with pickle).

**Comment by wangyinz** [comment]
Well, just found another more serious problem, which is pretty much expected - Spark relies on pickle to parallelize the data. I just did a quick test like this within the docker-compose environment:

```
from mspasspy import CoreTimeSeries
d=[CoreTimeSeries(100),CoreTimeSeries(99),CoreTimeSeries(98)]
data = spark.sparkContext.parallelize(d).map(lambda tr: (tr.s))
```

and it errors out with `TypeError: can't pickle mspasspy.CoreTimeSeries objects`. I guess we will have to implement pickle for sure.

**Comment by pavlis** [comment]
I made some important progress on this problem.   Can't say 100% it will all work, but it has worked on toy problems.  Hoping to do some real tests with mongodb later today, although I have some conflicts that may prevent me from completing the test.

The trick I learned is easily seen by this little python snippet (just typing this - no guarantee it will work if you can and paste it)

from mspasspy import Seismogram
import pickle
d=CoreSeismogram(100)  # make an empty Seismogram with 100 samples
ub=bytes(d.u)

pickle dump/load dumps/loads work on the bytearray ub.  (Note bytes is an immutable bytearray).  That apparently works because d.u in the wrappers are equivalenced with a byte array which bytes knows how to handle. 

For reading there is a numpy function (I don't remember the name - I think is decode) to turn the bytearray into doubles. Worst case have to write a loop to copy same into a d.u.   That will be SLOW, however, if it has to be done by the interpreter.   Let's remember and old axiom I learned many years ago:  make it work before you make it fast.

**Comment by wangyinz** [comment]
Yeah, that seems to be a viable workaround to make Spark functioning for now.

**Comment by pavlis** [comment]
In some ways this is a new topic, but it fits this category with some new twists.   It appears spark will use pickle under the hood to serialize any C++ objects it is given.  Hence, I've been working on building wrappers to do that in pybind11 for the past several days.   The status is this:
1.  Seismogram seems to be working fine.  It passes all my simple, but exhaustive tests.  
2.  TimeSeries wrappers compile and link, but something is broken because pickle.dumps is seg faulting on me.   I checked in this version and am going to try to debug this with totalview on the IU HPC systems.   Having trouble getting two interactive debuggers that use gdb to work right on ubuntu machine, so I'm punting to totalview for now.

Now with that background I wanted to preserve something.  Both the implementation for Seismogram and TimeSeries required me to resort to enforcing the mantra "make it work before you make it fast".   I could not figure out a means to build the tuple pybind11's interface for doing pickle required without copying the data arrays for both objects.   In the case of Seismogram I ended up using boost serialization to serialize the data matrix into a (long) text string.   I know from previous experience with the seispp library that text serialization of seismic data is a very large bottleneck, but it created a working prototype anyway.

TimeSeries presented a different problem.  I tried to figure out how to directly map the data array into what pybind11 calls a buffer object, but gave up and did a different copy operation.  In this case, I created a numpy array through the (poorly documented) array_t template in pybind11.  I copied the vector<double> container values into the numpy array, which could be pickled.  There is a high probability my seg fault is being created by not understanding that interface correctly.   

I'm writing this down right now to preserve something.  That is, I'm about 100% sure a faster, but more obscure solution to creating the pickle interface for both Seismogram and TimeSeries is to crack the (poorly documented) interaction between array_t and the idea of a buffer object.  As I read the documentation it seems fairly likely that with this mechanism we can create and alias that makes the data stored in the Seismogram and TimeSeries containers look like a numpy array that can be pickled directly.   i.e. we can create a symbol that is more or less a wrapper that uses the same memory for a fake numpy array as the data in the original container (Seismogram's u matrix or TimeSeries's s vector).  I am leaving this till later when we prove (likely) it is worth the effort.

**Comment by pavlis** [comment]
Update a few hours later.   I got the array_t feature to work, as best I can yet tell, with TimeSeries and the std:: container.   Digging into the include file for pybind11 I found that all that is required to fake pickle into thinking a contiguous array of doubles (required for std::vector container) stored in vector is a python array it took only this line:
```
py::array_t<double, py::array::f_style> darr(self.ns,&(self.s[0]));
```
where self.ns defines the length of the buffer and &(self.s[0]) is a standard trick to define a pointer to the first element of the vector<double> container used to hold data in the Seismogram object.  REALLY slick trick to reduce overhead.

I have not yet done the comparable trick for Seismogram.  Can and should be possible, but I have to figure out how to set the size and stride variables used by numpy and set in the buffer object used to fake python into thinking this it would be a numpy array.

**Comment by pavlis** [comment]
Looking at buffer_info details and pybind11 I am not at all convinced it would be worth the effort to try to use that approach to serialize the dmatrix in the Seismogram object.   When I consider how I'd have to implement it, both cases will end up having to convert to some version of a string representation of the binary data.   I would be surprised if the overhead in pickle's serialization of a numpy array is any faster than boost's serialization.   They are pretty much doing the same thing with different formats as best i can judge without digging a lot deeper.  Hence, let's stick with this form now.

**Comment by wangyinz** [comment]
Closing this as completed.

The core Python bindings now define pickle support for `TimeSeries`, `Seismogram`, ensembles, metadata/history objects, and relevant algorithms. The Python test suite contains explicit `pickle.dumps`/`pickle.loads` round-trip coverage for those objects, including distributed-use cases. That resolves the serialization concern raised here.
